File: packages/liveline-mobile/ios/HybridLivelineView.swift
import Foundation
import LivelineKit
import NitroModules
import UIKit

/// The Nitro HybridView implementation: instantiates the native `LivelineView`
/// from LivelineKit and forwards props/methods to it. No drawing or chart maths
/// live here — that all belongs to LivelineKit.
///
/// Types from LivelineKit are qualified because the Nitrogen-generated types
/// (`LivelinePoint`, `CandlePoint`, `LivelineMode`, `LivelineTheme`, …) share
/// their names.
///
/// Marked `@MainActor` — Nitro applies view props and calls methods on the main
/// thread, keeping it on the same actor as the `@MainActor` `LivelineView`.
/// (The binding pod builds in Swift 5 language mode; see the podspec/Podfile.)
@MainActor
final class HybridLivelineView: HybridLivelineSpec {
    private let chart = LivelineKit.LivelineView()
    private let bar = LivelineKit.WindowBarView()
    private let legend = LivelineKit.SeriesLegendView()
    private var lastValue: Double?

    /// Notifies JS when a window button is tapped (set by Nitro).
    var onWindowChange: ((_ secs: Double) -> Void)?
    /// Notifies JS when the mode toggle is tapped; its presence shows the toggle.
    var onModeChange: ((_ mode: LivelineMode) -> Void)? {
        didSet { bar.showModeToggle = onModeChange != nil }
    }

    private lazy var container: LivelineContainerView = {
        bar.isHidden = true
        bar.onSelect = { [weak self] secs in
            guard let self else { return }
            // Optimistic native update (no JS round-trip lag), then notify JS.
            self.chart.windowSeconds = secs
            self.bar.activeSecs = secs
            self.onWindowChange?(secs)
        }
        bar.onModeSelect = { [weak self] candle in
            guard let self else { return }
            self.chart.mode = candle ? .candle : .line
            self.bar.isCandle = candle
            self.onModeChange?(candle ? .candle : .line)
        }
        legend.isHidden = true
        legend.onToggle = { [weak self] id in self?.chart.toggleSeries(id) }
        return LivelineContainerView(legend: legend, bar: bar, chart: chart)
    }()

    /// The UIView Nitro mounts into the RN view tree — the bar stacked above the chart.
    var view: UIView { container }

    // MARK: Window bar

    var windows: [WindowOption]? {
        didSet {
            let ws = windows ?? []
            bar.windows = ws.map { (label: $0.label, secs: $0.secs) }
            bar.isHidden = ws.isEmpty
            container.setNeedsLayout()
        }
    }

    var windowStyle: LivelineWindowStyle? {
        didSet {
            switch windowStyle {
            case .rounded: bar.style = .rounded
            case .text: bar.style = .text
            default: bar.style = .default
            }
        }
    }

    // MARK: Data

    var data: [LivelinePoint]? {
        didSet {
            guard let data else { return }
            chart.setData(data.map { LivelineKit.LivelinePoint(time: $0.time, value: $0.value) })
        }
    }
    var series: [LivelineSeries]? {
        didSet {
            let list = series ?? []
            chart.setSeries(
                list.map { s in
                    LivelineView.SeriesInput(
                        id: s.id, color: UIColor(cssString: s.color), label: s.label,
                        data: s.data.map { LivelineKit.LivelinePoint(time: $0.time, value: $0.value) })
                })
            legend.items = list.map {
                .init(id: $0.id, color: UIColor(cssString: $0.color), label: $0.label ?? $0.id)
            }
            container.setNeedsLayout()
        }
    }

    var value: Double? {
        didSet {
            guard let value, value != lastValue else { return }
            lastValue = value
            chart.push(LivelineKit.LivelinePoint(time: Date().timeIntervalSince1970, value: value))
        }
    }

    // MARK: Candle mode

    var mode: LivelineMode? {
        didSet {
            chart.mode = (mode == .candle) ? .candle : .line
            bar.isCandle = mode == .candle
        }
    }

    var candles: [CandlePoint]? {
        didSet {
            chart.candles = (candles ?? []).map {
                LivelineKit.LivelineCandle(
                    time: $0.time, open: $0.open, high: $0.high, low: $0.low, close: $0.close)
            }
        }
    }

    var candleWidth: Double? {
        didSet { if let candleWidth { chart.candleWidth = candleWidth } }
    }

    var liveCandle: CandlePoint? {
        didSet {
            chart.liveCandle = liveCandle.map {
                LivelineKit.LivelineCandle(
                    time: $0.time, open: $0.open, high: $0.high, low: $0.low, close: $0.close)
            }
        }
    }

    // MARK: Appearance

    var color: String? {
        didSet { if let color { chart.color = UIColor(cssString: color) } }
    }

    var theme: LivelineTheme? {
        didSet {
            chart.theme = (theme == .light) ? .light : .dark
            bar.isDark = theme != .light
            legend.isDark = theme != .light
        }
    }

    var surfaceColor: String? {
        // Empty string (the JS default) means "no override" → theme default.
        didSet {
            if let surfaceColor, !surfaceColor.isEmpty {
                chart.surfaceColor = UIColor(cssString: surfaceColor)
            } else {
                chart.surfaceColor = nil
            }
        }
    }

    var lineWidth: Double? {
        didSet { if let lineWidth { chart.lineWidth = CGFloat(lineWidth) } }
    }

    // MARK: Time

    var window: Double? {
        didSet {
            if let window {
                chart.windowSeconds = window
                bar.activeSecs = window
            }
        }
    }

    // MARK: Feature flags

    var grid: Bool? {
        didSet { chart.grid = grid ?? true }
    }
    var badge: Bool? {
        didSet { chart.badge = badge ?? true }
    }
    var badgeTail: Bool? {
        didSet { chart.badgeTail = badgeTail ?? true }
    }
    var badgeVariant: LivelineBadgeVariant? {
        didSet {
            switch badgeVariant {
            case .minimal: chart.badgeVariant = .minimal
            case .accent: chart.badgeVariant = .accent
            default: chart.badgeVariant = .default
            }
        }
    }
    var momentum: LivelineMomentum? {
        didSet {
            switch momentum {
            case .off: chart.momentum = .off
            case .auto, .none: chart.momentum = .auto
            case .up: chart.momentum = .up
            case .down: chart.momentum = .down
            case .flat: chart.momentum = .flat
            }
        }
    }
    var fill: Bool? {
        didSet { chart.fill = fill ?? true }
    }
    var scrub: Bool? {
        didSet { chart.scrub = scrub ?? true }
    }
    var tooltipY: Double? {
        didSet { chart.tooltipY = tooltipY ?? 14 }
    }
    var tooltipOutline: Bool? {
        didSet { chart.tooltipOutline = tooltipOutline ?? true }
    }
    var padding: LivelinePadding? {
        didSet {
            chart.padTopOverride = padding?.top; chart.padRightOverride = padding?.right
            chart.padBottomOverride = padding?.bottom; chart.padLeftOverride = padding?.left
        }
    }
    // Legend isn't part of the container yet; stored until it is wired.
    var seriesToggleCompact: Bool?
    var onSeriesToggle: ((_ id: String, _ visible: Bool) -> Void)?
    var pulse: Bool? {
        didSet { chart.pulse = pulse ?? true }
    }
    var exaggerate: Bool? {
        didSet { chart.exaggerate = exaggerate ?? false }
    }
    var paused: Bool? {
        didSet { chart.paused = paused ?? false }
    }
    var loading: Bool? {
        didSet { chart.loading = loading ?? false }
    }
    var emptyText: String? {
        didSet { if let emptyText { chart.emptyText = emptyText } }
    }
    var showValue: Bool? {
        didSet { chart.showValue = showValue ?? false }
    }
    var valueMomentumColor: Bool? {
        didSet { chart.valueMomentumColor = valueMomentumColor ?? false }
    }
    var haptics: Bool? {
        didSet { chart.haptics = haptics ?? false }
    }
    var degen: Bool? {
        didSet { chart.degen = degen ?? false }
    }
    var lerpSpeed: Double? {
        didSet { if let lerpSpeed { chart.lerpSpeed = lerpSpeed } }
    }
    var referenceLine: LivelineReference? {
        // A non-finite value (the JS wrapper's sentinel) means "no line".
        didSet {
            if let r = referenceLine, r.value.isFinite {
                chart.referenceLine = LivelineKit.ReferenceLine(value: r.value, label: r.label)
            } else {
                chart.referenceLine = nil
            }
        }
    }

    var orderbook: LivelineOrderbook? {
        didSet { chart.setOrderbook(orderbook.map(Self.toOrderbook)) }
    }

    private static func toOrderbook(_ o: LivelineOrderbook) -> LivelineKit.OrderbookData {
        LivelineKit.OrderbookData(
            bids: o.bids.map { LivelineKit.OrderbookLevel(price: $0.price, size: $0.size) },
            asks: o.asks.map { LivelineKit.OrderbookLevel(price: $0.price, size: $0.size) })
    }

    // MARK: Declarative value formatting

    var valuePrefix: String? { didSet { rebuildFormatter() } }
    var valueSuffix: String? { didSet { rebuildFormatter() } }
    var valueDecimals: Double? { didSet { rebuildFormatter() } }
    var currency: String? { didSet { rebuildFormatter() } }
    var locale: String? {
        didSet {
            chart.formattingLocale = locale.flatMap { $0.isEmpty ? nil : Locale(identifier: $0) } ?? .current
            rebuildFormatter()
        }
    }
    var useGrouping: Bool? { didSet { rebuildFormatter() } }
    var fontFamily: String? {
        didSet {
            let f = (fontFamily?.isEmpty ?? true) ? nil : fontFamily
            chart.fontFamily = f
            bar.fontFamily = f
        }
    }

    private func rebuildFormatter() {
        let nf = NumberFormatter()
        nf.locale = locale.flatMap { $0.isEmpty ? nil : Locale(identifier: $0) } ?? .current
        nf.usesGroupingSeparator = useGrouping ?? true
        if let currency, !currency.isEmpty {
            // Currency style provides the symbol, its placement and the currency's
            // own default fraction digits (0 for JPY, 2 for USD, …); `valuePrefix`/
            // `valueSuffix`/`valueDecimals` don't apply here.
            nf.numberStyle = .currency
            nf.currencyCode = currency
            chart.formatValue = { nf.string(from: NSNumber(value: $0)) ?? "\($0)" }
        } else {
            nf.numberStyle = .decimal
            let decimals = Int(valueDecimals ?? 2)
            nf.minimumFractionDigits = decimals
            nf.maximumFractionDigits = decimals
            let prefix = valuePrefix ?? ""
            let suffix = valueSuffix ?? ""
            chart.formatValue = { prefix + (nf.string(from: NSNumber(value: $0)) ?? "\($0)") + suffix }
        }
    }

    // MARK: Methods

    func push(point: LivelinePoint, seriesId: String?) throws {
        // `push` is called from JS, which may not be the main thread. The chart
        // is UIKit/@MainActor and its render loop reads the buffer on main — so
        // hop to main to avoid a data race (intermittent SIGABRT deep in UIKit).
        let p = LivelineKit.LivelinePoint(time: point.time, value: point.value)
        if Thread.isMainThread {
            chart.push(p, seriesId: seriesId)
        } else {
            DispatchQueue.main.async { self.chart.push(p, seriesId: seriesId) }
        }
    }

    func pushOrderbook(orderbook: LivelineOrderbook) throws {
        // Like `push`, may be called off the main thread; hop to main.
        let book = Self.toOrderbook(orderbook)
        if Thread.isMainThread {
            chart.setOrderbook(book)
        } else {
            DispatchQueue.main.async { self.chart.setOrderbook(book) }
        }
    }
}

extension UIColor {
    /// Minimal CSS-colour parse for the binding: `#rgb` or `#rrggbb`. Falls back
    /// to the liveline default blue for anything else.
    fileprivate convenience init(cssString: String) {
        let s = cssString.trimmingCharacters(in: .whitespaces)
        if s.hasPrefix("#") {
            var hex = String(s.dropFirst())
            if hex.count == 3 { hex = hex.map { "\($0)\($0)" }.joined() }
            if hex.count == 6, let v = UInt64(hex, radix: 16) {
                self.init(
                    red: CGFloat((v & 0xFF0000) >> 16) / 255,
                    green: CGFloat((v & 0x00FF00) >> 8) / 255,
                    blue: CGFloat(v & 0x0000FF) / 255,
                    alpha: 1)
                return
            }
        }
        self.init(red: 59 / 255, green: 130 / 255, blue: 246 / 255, alpha: 1)
    }
}

/// Stacks the interval bar above the chart. When there are no windows the bar is
/// hidden and the chart fills the whole view.
final class LivelineContainerView: UIView {
    private let legend: LivelineKit.SeriesLegendView
    private let bar: LivelineKit.WindowBarView
    private let chart: LivelineKit.LivelineView

    init(
        legend: LivelineKit.SeriesLegendView, bar: LivelineKit.WindowBarView,
        chart: LivelineKit.LivelineView
    ) {
        self.legend = legend
        self.bar = bar
        self.chart = chart
        super.init(frame: .zero)
        addSubview(chart)
        addSubview(legend)  // multi-series legend at the top
        addSubview(bar)  // interval bar at the bottom
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) { fatalError("init(coder:) not supported") }

    override func layoutSubviews() {
        super.layoutSubviews()
        // Legend on top, interval bar at the bottom, chart fills between.
        let legendH = legend.isHidden ? 0 : legend.intrinsicContentSize.height
        let barH = bar.isHidden ? 0 : bar.intrinsicContentSize.height
        legend.frame = CGRect(x: 0, y: 0, width: bounds.width, height: legendH)
        chart.frame = CGRect(x: 0, y: legendH, width: bounds.width, height: bounds.height - legendH - barH)
        bar.frame = CGRect(x: 0, y: bounds.height - barH, width: bounds.width, height: barH)
    }
}

File: packages/liveline-mobile/android/src/main/java/com/margelo/nitro/liveline/HybridLivelineView.kt
package com.margelo.nitro.liveline

import android.content.Context
import android.graphics.Color
import android.graphics.Typeface
import android.os.Looper
import android.view.Gravity
import android.view.View
import android.widget.LinearLayout
import com.liveline.LivelineView
import com.liveline.SeriesLegendView
import com.liveline.WindowBarView
import com.liveline.core.BadgeVariant as CoreBadge
import com.liveline.core.LivelineCandle as CoreCandle
import com.liveline.core.LivelineMode as CoreMode
import com.liveline.core.LivelineTheme as CoreTheme
import com.liveline.core.Momentum as CoreMomentum
import com.liveline.core.OrderbookData as CoreBook
import com.liveline.core.OrderbookLevel as CoreLevel
import com.liveline.core.ReferenceLine as CoreRef

/**
 * The Android Nitro HybridView for Liveline: it maps the RN props/methods onto
 * the shared `com.liveline.LivelineView` renderer. Props arrive on the UI thread
 * via the ViewManager's Fabric `updateState`; the imperative methods (`push`,
 * `pushOrderbook`) are called from the JS thread, so they hop to main.
 */
class HybridLivelineView(context: Context) : HybridLivelineSpec() {
    companion object {
        init {
            // Load the C++ library before the super constructor wires the C++ part.
            LivelineMobileOnLoad.initializeNative()
        }
    }

    private val chart = LivelineView(context)
    private val windowBar = WindowBarView(context).apply { visibility = View.GONE }
    private val seriesLegend = SeriesLegendView(context).apply { visibility = View.GONE }
    // Vertical container: the legend (shown only for multi-series) on top, the
    // chart filling, the interval bar (shown only when `windows` is set) below —
    // mirroring the iOS container.
    private val container = LinearLayout(context).apply {
        orientation = LinearLayout.VERTICAL
        val dp = resources.displayMetrics.density
        addView(
            seriesLegend,
            LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT,
            ).apply { gravity = Gravity.CENTER_HORIZONTAL; bottomMargin = (8 * dp).toInt() },
        )
        addView(chart, LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, 0, 1f))
        addView(
            windowBar,
            LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT,
            ).apply { gravity = Gravity.CENTER_HORIZONTAL; topMargin = (8 * dp).toInt() },
        )
    }
    override val view: View get() = container
    private var lastValue: Double? = null

    init {
        windowBar.onSelect = { secs ->
            chart.windowSeconds = secs
            onWindowChange?.invoke(secs)
        }
        windowBar.onModeSelect = { candle ->
            chart.mode = if (candle) CoreMode.CANDLE else CoreMode.LINE
            onModeChange?.invoke(if (candle) LivelineMode.CANDLE else LivelineMode.LINE)
        }
        seriesLegend.onToggle = { id ->
            val visible = chart.toggleSeries(id)
            onSeriesToggle?.invoke(id, visible)
        }
    }

    private fun refreshLegend() {
        val info = chart.seriesInfo()
        if (info.isEmpty()) { seriesLegend.visibility = View.GONE; return }
        seriesLegend.visibility = View.VISIBLE
        seriesLegend.setItems(info.map { SeriesLegendView.Item(it.first, it.second, it.third) })
    }

    private fun refreshWindowBar() {
        val w = windows
        val show = !w.isNullOrEmpty() || modeToggle == true
        if (!show) { windowBar.visibility = View.GONE; return }
        windowBar.visibility = View.VISIBLE
        windowBar.showModeToggle = modeToggle == true
        if (!w.isNullOrEmpty()) windowBar.setWindows(w.map { WindowBarView.Window(it.label, it.secs) }, window ?: w.first().secs)
        else windowBar.setWindows(emptyList(), window ?: 0.0)
    }

    private fun corePoint(p: LivelinePoint) = com.liveline.core.LivelinePoint(p.time, p.value)
    private fun coreCandle(c: CandlePoint) = CoreCandle(c.time, c.open, c.high, c.low, c.close)
    private fun coreLevel(l: LivelineOrderbookLevel) = CoreLevel(l.price, l.size)
    private fun coreBook(b: LivelineOrderbook) =
        CoreBook(b.bids.map { coreLevel(it) }, b.asks.map { coreLevel(it) })

    private fun onMain(block: () -> Unit) {
        if (Looper.myLooper() == Looper.getMainLooper()) block() else chart.post(block)
    }

    private fun applyCandles() {
        val c = candles ?: return
        chart.setCandles(c.map { coreCandle(it) }, liveCandle?.let { coreCandle(it) }, candleWidth ?: 1.0)
    }

    /** Builds the value formatter from currency/locale/grouping/prefix/suffix/
     *  decimals, mirroring the iOS binding's NumberFormatter logic. */
    private fun rebuildFormatter() {
        val loc = locale?.takeIf { it.isNotEmpty() }?.let { java.util.Locale.forLanguageTag(it) }
            ?: java.util.Locale.getDefault()
        val cur = currency
        if (!cur.isNullOrEmpty()) {
            // Currency style provides the symbol/placement and the currency's own
            // fraction digits; prefix/suffix/decimals don't apply (matches iOS).
            val nf = java.text.NumberFormat.getCurrencyInstance(loc)
            runCatching { nf.currency = java.util.Currency.getInstance(cur) }
            nf.isGroupingUsed = useGrouping ?: true
            chart.formatValue = { nf.format(it) }
        } else {
            val nf = java.text.NumberFormat.getNumberInstance(loc)
            val d = (valueDecimals ?: 2.0).toInt()
            nf.minimumFractionDigits = d
            nf.maximumFractionDigits = d
            nf.isGroupingUsed = useGrouping ?: true
            val pre = valuePrefix ?: ""
            val suf = valueSuffix ?: ""
            chart.formatValue = { pre + nf.format(it) + suf }
        }
        chart.invalidate()
    }

    /** Resolves an RN font-family name to a Typeface (bundled fonts via
     *  ReactFontManager; falls back to a system family). */
    private fun resolveTypeface(family: String): Typeface? = runCatching {
        com.facebook.react.common.assets.ReactFontManager.getInstance()
            .getTypeface(family, Typeface.NORMAL, chart.context.assets)
    }.getOrElse { runCatching { Typeface.create(family, Typeface.NORMAL) }.getOrNull() }

    // ── Props ───────────────────────────────────────────────────────────────
    override var data: Array<LivelinePoint>? = null
        set(v) { field = v; if (v != null) chart.setData(v.map { corePoint(it) }) }
    override var series: Array<LivelineSeries>? = null
        set(v) {
            field = v
            if (v != null) chart.setSeries(v.map { s ->
                LivelineView.SeriesInput(
                    s.id,
                    runCatching { Color.parseColor(s.color) }.getOrDefault(chart.accent),
                    s.label,
                    s.data.map { corePoint(it) },
                )
            })
            refreshLegend()
        }
    override var value: Double? = null
        set(v) {
            field = v
            if (v != null && v != lastValue) {
                lastValue = v
                chart.push(com.liveline.core.LivelinePoint(System.currentTimeMillis() / 1000.0, v))
            }
        }
    override var mode: LivelineMode? = null
        set(v) {
            field = v
            chart.mode = if (v == LivelineMode.CANDLE) CoreMode.CANDLE else CoreMode.LINE
            windowBar.isCandle = v == LivelineMode.CANDLE
        }
    override var modeToggle: Boolean? = null
        set(v) { field = v; refreshWindowBar() }
    override var candles: Array<CandlePoint>? = null
        set(v) { field = v; applyCandles() }
    override var candleWidth: Double? = null
        set(v) { field = v; applyCandles() }
    override var liveCandle: CandlePoint? = null
        set(v) { field = v; applyCandles() }
    override var color: String? = null
        set(v) { field = v; if (!v.isNullOrEmpty()) runCatching { chart.accent = Color.parseColor(v) } }
    override var theme: LivelineTheme? = null
        set(v) {
            field = v
            chart.theme = if (v == LivelineTheme.LIGHT) CoreTheme.LIGHT else CoreTheme.DARK
            windowBar.isDark = v != LivelineTheme.LIGHT
            seriesLegend.isDark = v != LivelineTheme.LIGHT
        }
    override var surfaceColor: String? = null
        set(v) {
            field = v
            chart.surfaceColor = v?.takeIf { it.isNotEmpty() }?.let { runCatching { Color.parseColor(it) }.getOrNull() }
        }
    override var lineWidth: Double? = null
        set(v) { field = v; if (v != null && v > 0) chart.lineWidth = v }
    override var window: Double? = null
        set(v) { field = v; if (v != null && v > 0) { chart.windowSeconds = v; windowBar.setActive(v) } }
    override var windows: Array<WindowOption>? = null
        set(v) { field = v; refreshWindowBar() }
    // The Android interval bar has a single (pill) style for now; the `windowStyle`
    // variants (rounded/text) are honoured on iOS.
    override var windowStyle: LivelineWindowStyle? = null
    override var onWindowChange: ((secs: Double) -> Unit)? = null
    override var onModeChange: ((mode: LivelineMode) -> Unit)? = null
    override var grid: Boolean? = null
        set(v) { field = v; if (v != null) chart.grid = v }
    override var badge: Boolean? = null
        set(v) { field = v; if (v != null) chart.badge = v }
    override var badgeTail: Boolean? = null
        set(v) { field = v; if (v != null) chart.badgeTail = v }
    override var badgeVariant: LivelineBadgeVariant? = null
        set(v) {
            field = v
            chart.badgeVariant = when (v) {
                LivelineBadgeVariant.MINIMAL -> CoreBadge.MINIMAL
                LivelineBadgeVariant.ACCENT -> CoreBadge.ACCENT
                else -> CoreBadge.DEFAULT
            }
        }
    override var momentum: LivelineMomentum? = null
        set(v) {
            field = v
            chart.momentum = when (v) {
                LivelineMomentum.OFF -> CoreMomentum.OFF
                LivelineMomentum.UP -> CoreMomentum.UP
                LivelineMomentum.DOWN -> CoreMomentum.DOWN
                LivelineMomentum.FLAT -> CoreMomentum.FLAT
                else -> CoreMomentum.AUTO
            }
        }
    override var fill: Boolean? = null
        set(v) { field = v; if (v != null) chart.fill = v }
    override var scrub: Boolean? = null
        set(v) { field = v; if (v != null) chart.scrub = v }
    override var tooltipY: Double? = null
        set(v) { field = v; if (v != null) chart.tooltipY = v }
    override var tooltipOutline: Boolean? = null
        set(v) { field = v; if (v != null) chart.tooltipOutline = v }
    override var padding: LivelinePadding? = null
        set(v) {
            field = v
            chart.padTopOverride = v?.top; chart.padRightOverride = v?.right
            chart.padBottomOverride = v?.bottom; chart.padLeftOverride = v?.left
        }
    override var seriesToggleCompact: Boolean? = null
        set(v) { field = v; if (v != null) seriesLegend.compact = v }
    override var onSeriesToggle: ((id: String, visible: Boolean) -> Unit)? = null
    override var pulse: Boolean? = null
        set(v) { field = v; if (v != null) chart.pulse = v }
    override var exaggerate: Boolean? = null
        set(v) { field = v; if (v != null) chart.exaggerate = v }
    override var paused: Boolean? = null
        set(v) { field = v; if (v != null) chart.paused = v }
    override var loading: Boolean? = null
        set(v) { field = v; if (v != null) chart.loading = v }
    override var emptyText: String? = null
        set(v) { field = v; if (v != null) chart.emptyText = v }
    override var showValue: Boolean? = null
        set(v) { field = v; if (v != null) chart.showValue = v }
    override var valueMomentumColor: Boolean? = null
        set(v) { field = v; if (v != null) chart.valueMomentumColor = v }
    override var haptics: Boolean? = null
        set(v) { field = v; if (v != null) chart.haptics = v }
    override var degen: Boolean? = null
        set(v) { field = v; if (v != null) chart.degen = v }
    override var lerpSpeed: Double? = null
        set(v) { field = v; if (v != null && v > 0) chart.lerpSpeed = v }
    override var referenceLine: LivelineReference? = null
        set(v) { field = v; chart.referenceLine = v?.let { CoreRef(it.value, it.label) } }
    override var orderbook: LivelineOrderbook? = null
        set(v) { field = v; chart.orderbook = v?.let { coreBook(it) } }
    override var valuePrefix: String? = null
        set(v) { field = v; chart.valuePrefix = v ?: ""; rebuildFormatter() }
    override var valueSuffix: String? = null
        set(v) { field = v; chart.valueSuffix = v ?: ""; rebuildFormatter() }
    override var valueDecimals: Double? = null
        set(v) { field = v; if (v != null) chart.valueDecimals = v.toInt(); rebuildFormatter() }
    override var currency: String? = null
        set(v) { field = v; rebuildFormatter() }
    override var locale: String? = null
        set(v) { field = v; rebuildFormatter() }
    override var useGrouping: Boolean? = null
        set(v) { field = v; rebuildFormatter() }
    override var fontFamily: String? = null
        set(v) {
            field = v
            chart.numberTypeface = v?.takeIf { it.isNotEmpty() }?.let { resolveTypeface(it) }
        }

    // ── Methods ─────────────────────────────────────────────────────────────
    override fun push(point: LivelinePoint, seriesId: String?) {
        val p = corePoint(point)
        onMain { if (seriesId != null) chart.pushSeries(seriesId, p) else chart.push(p) }
    }

    override fun pushOrderbook(orderbook: LivelineOrderbook) {
        val b = coreBook(orderbook)
        onMain { chart.orderbook = b }
    }
}
